pkg/mimirtool/commands/analyse_prometheus.go

```diff
@@ -34,173 +38,211 @@ type PrometheusAnalyzeCommand struct {
-		grafanaMetrics                     = analyze.MetricsInGrafana{}
-		rulerMetrics                       = analyze.MetricsInRuler{}
-		metricsUsed                        []string
+		metricsUsed    model.LabelValues
+		grafanaMetrics = &analyze.MetricsInGrafana{}
+		rulerMetrics   = &analyze.MetricsInRuler{}
 	)
 
-	if _, err := os.Stat(cmd.grafanaMetricsFile); err == nil {
-		hasGrafanaMetrics = true
-		byt, err := os.ReadFile(cmd.grafanaMetricsFile)
-		if err != nil {
-			return err
-		}
-		if err := json.Unmarshal(byt, &grafanaMetrics); err != nil {
-			return err
-		}
-		metricsUsed = append(metricsUsed, grafanaMetrics.MetricsUsed...)
+	if err := parseMetricFileIfExist(cmd.grafanaMetricsFile, grafanaMetrics); err != nil {
+		return nil, err
 	}
+	metricsUsed = append(metricsUsed, grafanaMetrics.MetricsUsed...)
 
-	if _, err := os.Stat(cmd.rulerMetricsFile); err == nil {
-		hasRulerMetrics = true
-		byt, err := os.ReadFile(cmd.rulerMetricsFile)
-		if err != nil {
-			return err
-		}
-		if err := json.Unmarshal(byt, &rulerMetrics); err != nil {
-			return err
-		}
-		metricsUsed = append(metricsUsed, rulerMetrics.MetricsUsed...)
+	if err := parseMetricFileIfExist(cmd.rulerMetricsFile, rulerMetrics); err != nil {
+		return nil, err
 	}
+	metricsUsed = append(metricsUsed, rulerMetrics.MetricsUsed...)
 
-	if !hasGrafanaMetrics && !hasRulerMetrics {
-		return errors.New("No Grafana or Ruler metrics files")
+	if len(metricsUsed) == 0 {
+		return nil, errors.New("no Grafana or Ruler metrics files")
 	}
 
+	return metricsUsed, nil
+}
+
+func (cmd *PrometheusAnalyzeCommand) newAPI() (v1.API, error) {
 	rt := api.DefaultRoundTripper
 	if cmd.username != "" {
 		rt = config.NewBasicAuthRoundTripper(cmd.username, config.Secret(cmd.password), "", api.DefaultRoundTripper)
 	}
-	promClient, err := api.NewClient(api.Config{
+
+	client, err := api.NewClient(api.Config{
 		Address:      cmd.address,
 		RoundTripper: rt,
 	})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	v1api := v1.NewAPI(promClient)
+	return v1.NewAPI(client), nil
+}
 
+func (cmd *PrometheusAnalyzeCommand) queryMetricNames(api v1.API) (model.LabelValues, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), cmd.readTimeout)
 	defer cancel()
-	metricNames, _, err := v1api.LabelValues(ctx, labels.MetricName, nil, time.Now().Add(-10*time.Minute), time.Now())
+
+	var metricNames model.LabelValues
+	err := withBackoff(ctx, func() error {
+		var err error
+		metricNames, _, err = api.LabelValues(ctx, labels.MetricName, nil, time.Now().Add(-10*time.Minute), time.Now())
+		return err
+	})
 	if err != nil {
-		return errors.Wrap(err, "error querying for metric names")
+		return nil, errors.Wrap(err, "error querying for metric names")
 	}
+
 	log.Infof("Found %d metric names\n", len(metricNames))
+	return metricNames, nil
+}
+
+func (cmd *PrometheusAnalyzeCommand) analyze(api v1.API, metrics model.LabelValues, filter func(model.LabelValue) bool) (analyseResult, error) {
```

Comment:
Actually, we respect the previous condition, we can not check the length since the `stats[metric]` is a struct: https://github.com/grafana/mimir/blob/b9a7d2b87827e09c2d1f8ae6f42ba025e9287f46/pkg/mimirtool/commands/analyse_prometheus.go#L151-L154

Reply:
I was thinking about this

https://github.com/grafana/mimir/blob/b9a7d2b87827e09c2d1f8ae6f42ba025e9287f46/pkg/mimirtool/commands/analyse_prometheus.go#L169-L177

vs 

https://github.com/anilmisirlioglu/mimir/blob/38d47bc16bd41e9abcfa478938db2246d04fc070/pkg/mimirtool/commands/analyse_prometheus.go#L175-L179


---

you can do 
```go
		return inUseMetricsAnalysisResult.stats[metric].totalCount > 0
```

Reply:
actually changing that return wouldn't retain the old behaviour. But I can't think of a reason why the query wouldn't return any results (i.e. empty vector), but we would still want to count the metric as used. Could the old behaviour be incorrect?

Reply:
You are right, we changed this line